### PR TITLE
expanding support for more complex types

### PR DESCRIPTION
**Pull request by jcal-15**
I was not planning on making this change so large. It started out as updating the list support to be more like the dict support.

Here is an attempt to summarize the main changes
1) refactored the code into several more functions
   - This allows for better handling of nested types
   1) An inner 'from_dict' function **UPDATE:** Done in #9
      - Allows a single NamespaceTypes instance to be created and used for the whole call
      - if a nested 'from_dict' is called called the given_args parameter passed in is not not copied or modified
    2) Adding 'handle list' function
       - Similar to the 'handle dict' function 
    3) Adding 'handle union' function
       - A common handler for union type. 
       - Called when a argument's type is a union or an inner type is a union (example a list's element are are union)
    4) Addin 'handle item' function
       - Called when we are looping over a list's elements or a dict's values. 
       - Based off of what type the value is, it will call another handler
2) more detailed location where a type check failed **UPDATE:** moved to #9
3) Checks for data-classes and attr classes for make the code faster in those cases. 
   - These type of classes are what I expect are the most commonly used when by clients to call this function.
4) Adding caching to ```resolve_str_forward_ref()```
   - This did not have as big of an impacted as I hoped it would have
   - It does make the call faster but they were pretty fast to begin with. 
5) All the type checking is only done at the end of each argument **UPDATE:** moved to #9
   - Removed it from the handle_dictionary function
6) cleanup of several type hint caught when I enable some type checking in vscode **UPDATE:** moved to #12
7) Cleanup of several tests because Python 3.6 is not supported. **UPDATE:** moved to #10

### CODE REVIEW

**Comment by jcal-15** [comment]
# Example of the changes to error reporting

## Source
```python
def test_error_message():
    @dataclass(frozen=True)
    class Class1:
        b: str

    @dataclass(frozen=True)
    class Class2:
        a: dict[str, list[Class1]]
        c: Optional['Class2']

    from_dict(Class2, a={"k":[0xbad]}, c=None, fd_check_types=True, fd_local_ns=locals())
```

## old error
```
Traceback (most recent call last):
  File ".\from_dict\_from_dict.py", line 285, in _from_dict_inner
    type_check(argument_value, cls_argument_type)
  File ".\from_dict\_from_dict.py", line 125, in type_check
    type_check(val, type_args[1])
  File ".\from_dict\_from_dict.py", line 117, in type_check
    type_check(element, targ)
  File ".\from_dict\_from_dict.py", line 93, in type_check
    raise FromDictTypeError([], t, type(v))
from_dict._from_dict.FromDictTypeError: For "", expected <class '__main__.test_error_message.<locals>.Class1'> but found <class 'int'> 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".\venv\man_test.py", line 170, in <module>
    test_error_message()
  File ".\venv\man_test.py", line 127, in test_error_message
    from_dict(Class2, a={"k":[0xbad]}, c=None, fd_check_types=True, fd_local_ns=locals())
  File ".\from_dict\_from_dict.py", line 212, in from_dict
    return _from_dict_inner(cls, given_args, fd_check_types, fd_copy_unknown, ns_types)
  File ".\from_dict\_from_dict.py", line 288, in _from_dict_inner
    raise FromDictTypeError(
from_dict._from_dict.FromDictTypeError: For "a", expected <class '__main__.test_error_message.<locals>.Class1'> but found <class 'int'>
```

## New Error
```
Traceback (most recent call last):
  File ".\venv\man_test.py", line 170, in <module>
    test_error_message()
  File ".\venv\man_test.py", line 127, in test_error_message
    from_dict(Class2, a={"k":[0xbad]}, c=None, fd_check_types=True, fd_local_ns=locals())
  File ".\from_dict\_from_dict.py", line 213, in from_dict
    return _from_dict_inner(cls, given_args, fd_check_types, fd_copy_unknown, ns_types)
  File ".\from_dict\_from_dict.py", line 266, in _from_dict_inner
    type_check([cls_argument_name], argument_value, cls_argument_type)
  File ".\from_dict\_from_dict.py", line 125, in type_check
    type_check(check_stack + [f"[{k!r}]"], val, type_args[1])
  File ".\from_dict\_from_dict.py", line 117, in type_check
    type_check(check_stack + [f"[{i}]"], element, targ)
  File ".\from_dict\_from_dict.py", line 93, in type_check
    raise FromDictTypeError(location(), t, type(v))
from_dict._from_dict.FromDictTypeError: For "a['k'][0]", expected <class '__main__.test_error_message.<locals>.Class1'> but found <class 'int'>
```

There is more I want to do with error handling but these pull request is already pretty big (too big..)

File: from_dict/_from_dict.py [hunk 1]
```diff
@@ -63,76 +62,86 @@ def __eq__(self, o: object) -> bool:
...
 
 if IS_GE_PYTHON38:
-
-    def get_origin(t):
-        return typing.get_origin(t)
-
-    def get_args(t):
-        return typing.get_args(t)
-
+    from typing import get_origin, get_args
```
Comment (wchresta) [inline from_dict/_from_dict.py:65]: Great improvement! Would you mind separating these improvements, which do not directly have to do with the change, into their own commit? That would reviewing the change a lot easier :).

File: from_dict/_from_dict.py [hunk 2]
```diff
@@ -63,76 +62,86 @@ def __eq__(self, o: object) -> bool:
...
         else:
-            return None
+            return ()
 
 
-def type_check(v, t) -> None:
+def type_check(check_stack: list, v: Any, t: type) -> None:
     """Raise RuntimeTypeError if given value does not agree with given type"""
     # This uses typing.get_args and typing.get_origin
+    def location():
```
Comment (wchresta) [inline from_dict/_from_dict.py:84]: Same here, it seems there's an improvement to error messages and a functional change. Could we separate those into separate commits?

File: from_dict/_from_dict.py (hunk 1 above)
Comment (jcal-15) [reply from_dict/_from_dict.py:65]: Yes, I can do that. I was thinking about that after I started the pull request.

Comment (jcal-15) [reply from_dict/_from_dict.py:65]: Okay, it is done. Hopefully I did not split them out into too may pull requests for you.
Besides the error location one, the others done have any changes that are client visible. 

After the other pull requests have been merged I will merge with main again and the PR should only show the core changes.

Comment (jcal-15) [reply from_dict/_from_dict.py:65]: @wchresta, I split out some of the changes into other pull requests. Are there other changes you want split out?

**Comment by jcal-15** [comment]
@wchresta, do you want to review these changes after I have split out some of them to other PRs?

File: tests/test_type_checking.py [hunk 3]
```diff
@@ -0,0 +1,584 @@
+import copy
```
Comment (wchresta) [inline tests/test_type_checking.py:1]: Nice!

**Review by wchresta — APPROVED** [review]
